File: ruby/phone-number/phone_number.rb
class PhoneNumber
  attr_reader :number

  def initialize(input)
    @number = clean(input)
  end

  def area_code
    number.slice(0, 3)
  end

  def prefix
    number.slice(3, 3)
  end

  def line_number
    number.slice(6, 4)
  end

  def to_s
    "(#{area_code}) #{prefix}-#{line_number}"
  end

  private

  def clean(input)
    num = input.scan(/\d/)
    num.shift if num.size == 11 && num.first == '1'
    return '0000000000' unless num.size == 10 && input.scan(/[a-z]/i).empty?
    num.join
  end
end

File: ecmascript/phone-number/phone-number.js
class PhoneNumber {
  constructor(num) {
    const digits = num.match(/\d/g).join('');
    if (digits.length == 10) {
      this.num = digits;
    } else if (digits.length == 11 && digits.startsWith('1')) {
      this.num = digits.slice(1);
    } else {
      this.num = '0000000000';
    }
  }

  number() {
    return this.num;
  }

  areaCode() {
    return this.num.slice(0, 3);
  }

  toString() {
    return `(${this.areaCode()}) ${this.num.slice(3, 6)}-${this.num.slice(6, 10)}`;
  }
}

export default PhoneNumber;
